mewarpx/tests/test_utils_pulsing.py

```diff
@@ -0,0 +1,151 @@
+
+def test_utils_pulsing_sim():
+    name = "pulsing_utility_test"
+    dim = 2
+
+    # Initialize and import only when we know dimension
+    mwxutil.init_libwarpx(ndim=dim, rz=False)
+    from mewarpx.mwxrun import mwxrun
+    from mewarpx.poisson_pseudo_1d import PoissonSolverPseudo1D
+    from mewarpx import assemblies, mepicmi
+    from mewarpx.utils_store import testing_util, pulsing
+
+    # Include a random run number to allow parallel runs to not collide. Using
+    # python randint prevents collisions due to numpy rseed below
+    testing_util.initialize_testingdir(name)
+
+    # Initialize each run with consistent, randomly-chosen, rseed. Use a random
+    # seed instead for initial dataframe generation.
+    # np.random.seed()
+    np.random.seed(9216024)
+
+    nx = 8
+    nz = 128
+    D_CA = 1e-4 # m
+
+    xmin = 0.0
+    zmin = 0.0
+
+    xmax = D_CA / nz * nx
+    zmax = D_CA
+
+    max_steps = 50
+
+    DT = 1e-10
+
+    #################################
+    # pulse parameters
+    ################################
+
+    V_off = -1.0
+    V_on = 7.5
+    pulse_period = 20e-9
+    pulse_length = 2e-9
+    t_rise = 1e-9
+    t_fall = 1e-9
+    wait_time = 0.5e-9
+
+    #####################################
+    # grid, solver and timesteps
+    #####################################
+
+    mwxrun.init_grid(xmin, xmax, zmin, zmax, nx, nz, max_grid_size=1)
+
+    solver = PoissonSolverPseudo1D(grid=mwxrun.grid)
+
+    mwxrun.simulation.solver = solver
+    mwxrun.init_timestep(30, DT=DT)
+    mwxrun.simulation.max_steps = max_steps
+
+    pulse_expr = pulsing.linear_pulse_function(
+        V_off=V_off, V_on=V_on, pulse_period=pulse_period,
+        pulse_length=pulse_length, t_rise=t_rise, t_fall=t_fall,
+        wait_time=wait_time
+    )
+    anode = assemblies.Anode(D_CA, pulse_expr, 100, 3.0)
+
+    electrons = mepicmi.Species(
+        particle_type='electron',
+        name='electrons',
+    )
+
+    # mwxrun.simulation.write_input_file()
+    # exit()
+
+    #################################
+    # simulation setup
+    ################################
+
+    mwxrun.init_run()
+
+    ################################
+    # Plot pulse function
+    ################################
+
+    def get_voltage(self, times):
+        voltages = np.zeros_like(times)
+        for ii, t in enumerate(times):
+            voltages[ii] = mwxrun.eval_expression_t(pulse_expr, t)
+        return voltages
+
+    pulse_function = type('testclass', (object,),
+        {
+            'get_voltage': get_voltage,
+            'V_off': V_off, 'V_on': V_on, 'pulse_period': pulse_period,
+            'pulse_length': pulse_length, 't_rise': t_rise,
+            't_fall': t_fall, 'wait_time': wait_time
+        }
+    )()
```

Comment:
It makes more sense to me to pull all this out into a separate class in `pulsing.py`. You could even make `plot_pulse` and `linear_pulse_function` member functions at that point. But having the plotting interface easily accessible by the user makes more sense to me than this syntax...